### PR TITLE
Try to get the shutdown and handling of exceptions right

### PR DESCRIPTION
**Pull request by knitori**
This doesn't really feel right though.

Some dirty hacks like line 77 in network.py …

I think there still are issues. Ctrl+C doesn't seem to work right yet.

### CODE REVIEW

File: shanghai/connection.py [hunk 1]
```diff
@@ -24,6 +24,9 @@ def writeline(self, line: bytes):
         self.writer.write(line)
         self.writer.write(b'\r\n')
 
+    def force_close(self):
+        self.writer.close()
+
     async def close(self):
         await self.writer.drain()
```
Comment (FichteFoll) [inline shanghai/connection.py]: `BaseTransport.close(self)`
> Close the transport. If the transport has a buffer for outgoing data, buffered data will be flushed asynchronously.

Meaning, we can ditch `.drain()` and force_close and close become equivalent.

File: shanghai/main.py [hunk 2]
```diff
@@ -11,15 +13,15 @@
...
     f = io.StringIO()
     print("exception_handler context:", file=f)
     pprint(context, stream=f)
+    traceback.print_exc(file=f)
     if 'task' in context:
         context['task'].print_stack(file=f)
     elif 'future' in context:
         context['future'].print_stack(file=f)
-    current_logger.error(f.getvalue())
+    # print(f.getvalue())
```
Comment (FichteFoll) [inline shanghai/main.py]: The entire value of `f` goes to waste like this?

File: shanghai/core.py [hunk 3]
```diff
@@ -19,4 +19,9 @@ def init_networks(self):
                 task=network_task,
                 network=network,
             )
+            network.mainloop_task = network_task
```
Comment (FichteFoll) [inline shanghai/core.py]: What is this for?

File: shanghai/network.py [hunk 4]
```diff
@@ -68,38 +69,52 @@ def next_server(self):
     async def run(self):
         self.log_context.push()
 
-        def cancel_other_task_if_failed(task, other_task):
-            current_logger.info('cancel_other_task_if_failed', task)
-            if task.exception():
-                task.print_stack()
+        def cancel_other(task: asyncio.Task, other_task: asyncio.Task):
+            try:
```
Comment (FichteFoll) [inline shanghai/network.py]: Could also check for `.cancelled()` instead.

File: shanghai/network.py [hunk 5]
```diff
@@ -167,6 +182,9 @@ def start_register(self):
             # context = Context(event, self)
             if event.name == 'message':
                 message = event.value
+                if message.command == 'PRIVMSG':
```
Comment (FichteFoll) [inline shanghai/network.py]: For now, this can stay. Imo

File: shanghai/main.py (hunk 2 above)
Comment (knitori) [inline shanghai/main.py]: yea. I had problems distinguishing exceptions that got handled by the function above and exceptions that didn't get handled. So I temporarily stopped the handler outputting the handled exceptions, to focus on the others.

File: shanghai/core.py (hunk 3 above)
Comment (knitori) [inline shanghai/core.py]: ´Network.shutdown´

File: shanghai/network.py (hunk 4 above)
Comment (knitori) [inline shanghai/network.py]: You're right. I realized only later that task.exception() does not raise an exception, if the task finished normally.

**Pull request closed without merging**